src/gate/index.js

```diff
@@ -54,21 +57,46 @@ Object.assign(UserGate.prototype, {
     return this._matchesSample(user) || this._matchesList(user);
   },
 
+  allowsUniform: function(user) {
+    return (this._matchesUniformSample(user) || this._matchesList(user));
+  },
+
   _matchesList: function(user) {
     if (!this._list) return false;
 
     return this._list.contains(user);
   },
 
   _matchesSample: function(user) {
-    if (!this._sample) return false;
+    if (!this._sample || !((this._sample >= 0) && (this._sample <= 1))) return false;
 
     // See if the user begins with a character in the sample set.
-    var effectiveCharacterSet = this._sampleCharacterSet.slice(
+    const effectiveCharacterSet = this._sampleCharacterSet.slice(
       0, Math.round(this._sampleCharacterSet.length * this._sample));
 
-    var userMatches = new RegExp('^[' + effectiveCharacterSet + ']', 'i').test(user);
+    const userMatches = new RegExp('^[' + effectiveCharacterSet + ']', 'i').test(user);
     return userMatches;
+  },
+
+  _matchesUniformSample: function(user) {
+    if (!this._sample || !((this._sample >= 0) && (this._sample <= 1))) return false;
+
+    // We've got to project `user` onto the sample space (i.e. convert it to a number between 0 and
+    // 1) in a way that is a) deterministic b) uniform. We do this by hashing the user, converting it
+    // to an unsigned 32-bit integer, and then normalizing by the size of the max UInt32--essentially
+    // this solution https://stats.stackexchange.com/a/70884 but with only two buckets ("in trial",
+    // or "not in trial").
+    const hash = new sha256();
+    const buf = hash.update(user).digest();
+
+    // This method of conversion-to-integer will truncate the hash to the first four bytes. But hashes
+    // are designed to be random in each individual bit https://crypto.stackexchange.com/a/26859, so
+    // this is no less uniform--just with greater likelihood of collisions, which is fine for us since
+    // we're collapsing the distribution onto two buckets anyway.
+    const uint32 = Buffer.from(buf).readUInt32BE(0);
```

Comment:
To maintain compatibility with our usage of `DataView`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/getUint32#Parameters

Reply:
Ohh good pickup. I thought that `DataView`'s [description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) saying "without having to care about the platform's endianness", and the `getUint32` function not having any suffix like `Buffer`'s do here, meant that `DataView` internally figured out endianness. Guess I should have read more closely! (And, it might have been a better choice to create different methods for big/little-endian, like Node did.)